### PR TITLE
Add unit tests for the layout engine

### PR DESCRIPTION
**Pull request by Adlai-Holler**
This adds a testing platform and some unit tests to the layout engine. Right now there are flags to toggle between the current behavior and the desired behavior, and they're set to "current behavior."

Issues reproduced (need to merge with GitHub issues):
- Extraneous `calculatedLayoutDidChange` calls. Not a biggie.
- If a layout pass comes in mid-transition, we will calculateLayout: for some nodes against their current bounds on main, but discard the result. https://github.com/TextureGroup/Texture/pull/424/files#diff-b1441b7e61c34b6f132da6c94bb10c0cR125
- If we do `setNeedsLayout` and get a layout pass while a transition is measuring, the transition's "to-layout" will win, but the setNeedsLayout+layoutPass should actually win since its data is newer. https://github.com/TextureGroup/Texture/pull/424/files#diff-b1441b7e61c34b6f132da6c94bb10c0cR204
- If we setNeedsLayout on a leaf node that is subject to multipass measurement (e.g. flexShrink) we will not update the layout for the leaf or the stack. We will remeasure it only against the fixed (second-pass) constrained size, and inevitably get a result confirming that the current size is correct, but we should have offered the node another unconstrained measurement pass. In my estimation this is the most serious issue. https://github.com/TextureGroup/Texture/pull/424/files#diff-b1441b7e61c34b6f132da6c94bb10c0cR284

I'd like to reproduce the orphan-node issue but I'm not sure what conditions would trigger it @nguyenhuy. We've got enough flexibility in the testing platform – I believe – to test an incredible variety of different scenarios.

PROTIP: You can take any expectation and add `.andDebugBreak()` to get a breakpoint when that call happens. So if you expect an async layout: `OCMExpect([nodeA.mock calculateLayoutThatFits:]).offMainThread()` you can change that to `OCMExpect([nodeA.mock calculateLayoutThatFits:]).offMainThread().andDebugBreak()` and when that calculateLayoutThatFits invocation happens, the debugger will pause for you. ✨Magic✨

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please ensure license is correct for debugbreak.h:</p>
<pre><code>//
//  debugbreak.h
//  Texture
//
//  Copyright (c) 2017-present, Pinterest, Inc.  All rights reserved.
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
</code></</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Tests/ASLayoutEngineTests.mm [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+  // nodeE
+  [fixture addSizeRange:{{0, 0}, {INFINITY, 1}} forNode:nodeE];
+  auto layoutE = [ASLayout layoutWithLayoutElement:nodeE size:{1,1} position:{9,0} sublayouts:nil];
+
+  [fixture addSizeRange:{{10, 1}, {10, 1}} forNode:nodeA];
+  auto layoutA = [ASLayout layoutWithLayoutElement:nodeA size:{10,1} position:ASPointNull sublayouts:@[ layoutB, layoutD, layoutE ]];
+  fixture.layout = layoutA;
+
+  ASLayoutSpecBlock specBlockA = ^ASLayoutSpec * _Nonnull(__kindof ASDisplayNode * _Nonnull node, ASSizeRange constrainedSize) {
+    return [ASStackLayoutSpec stackLayoutSpecWithDirection:ASStackLayoutDirectionHorizontal spacing:0 justifyContent:ASStackLayoutJustifyContentSpaceBetween alignItems:ASStackLayoutAlignItemsStart children:@[ nodeB, nodeC, nodeE ]];
```
Comment (nguyenhuy) [inline Tests/ASLayoutEngineTests.mm]: Shouldn't the `children` contains nodes B, D, E instead of B, C, E?

File: Tests/ASLayoutEngineTests.mm [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+ * From fixture 1:
+ *   B survives with same layout
+ *   C is removed
+ *   D survives with new layout
+ *   E joins with first layout
+ *
+ * From fixture 2:
+ *   B survives with same layout
+ *   C is removed
+ *   D is removed
```
Comment (nguyenhuy) [inline Tests/ASLayoutEngineTests.mm]: Should be "D joins the first layout" since it's a new subnode.

File: Tests/ASLayoutEngineTests.mm [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+          });
+        });
+      });
+    }];
+  }
+
+  // The code in this section is designed to move in time order, all on the main thread:
+
+  OCMExpect([nodeA.mock animateLayoutTransition:OCMOCK_ANY]).onMainThread();
+  OCMExpect([nodeA.mock didCompleteLayoutTransition:OCMOCK_ANY]).onMainThread();
```
Comment (nguyenhuy) [inline Tests/ASLayoutEngineTests.mm]: These method calls are only expected when the correct behavior is **not** enforced.  The correct behavior is that the layout transition will be cancelled and this it won't animate and complete.

File: Tests/ASLayoutEngineTests.mm [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+  [self verifyFixture:fixture2];
+}
+
+/**
+ * Start at fixture 1.
+ * Trigger an async transition to fixture 2.
+ * While it's measuring, on main switch to fixture 4 (setNeedsLayout A, D) and run a CA layout pass.
+ *
+ * Correct behavior, we end up at fixture 4 since it's newer.
+ * Current incorrect behavior, we end up at fixture 2 and we remeasure surviving node C.
```
Comment (nguyenhuy) [inline Tests/ASLayoutEngineTests.mm]: I believe this unexpected behavior is caused by [this early check]( https://github.com/TextureGroup/Texture/blob/master/Source/ASDisplayNode.mm#L920) in `__layout`, which is introduced in [#2657](https://github.com/facebookarchive/AsyncDisplayKit/pull/2657/files#diff-230ee3845413c2a11bf13fbdb7823ddfR1581).

File: Tests/Common/debugbreak.h [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+//  debugbreak.h
+//  Texture
+//
+//  Copyright (c) 2017-present, Pinterest, Inc.  All rights reserved.
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
```
Comment (nguyenhuy) [inline Tests/Common/debugbreak.h]: Do we really need this license? Or added just to make Danger happy?

File: Tests/ASLayoutEngineTests.mm [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+
+    [window layoutIfNeeded];
+    [self verifyFixture:fixture1];
+  } else {
+    /*
+     * Incorrect behavior: nodeB is remeasured against the second (fixed-width) constraint.
+     * The returned value (8) is clamped to the fixed with (7), and then compared to the previous
+     * width (7) and we decide not to propagate up the invalidation, and we stay stuck on the old
+     * layout (fixture3).
+     */
```
Comment (nguyenhuy) [inline Tests/ASLayoutEngineTests.mm]: Is this because nodes only store their last-used constraint?

**Comment by nguyenhuy** [comment]
Btw, I don't think this diff is trivial.  I'd suggest to add something to the change log.  It's your call though.

File: Tests/ASLayoutEngineTests.mm (hunk 6 above)
Comment (Adlai-Holler) [inline Tests/ASLayoutEngineTests.mm]: Yes exactly!

File: Tests/Common/debugbreak.h (hunk 5 above)
Comment (Adlai-Holler) [inline Tests/Common/debugbreak.h]: Just to make Danger happy.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/d7a4f918d4737900c8717f8eeb4c2b7d/raw/74d1b874d1702f8c1f4156043126fa7a9078a8cf/log.txt)

**Comment by Adlai-Holler** [comment]
The CI failure this time was interesting! I'll put this info somewhere else in the future, but I want to make sure not to lose it:

This is an extremely idealized version of a 3 hour slog to diagnose a very tricky (for me) race condition. I believe next time the slog can be way shorter with these learnings. To diagnose, I override `-invokeTest` like below, and told Xcode just to run the single test method (right click it, or use Ctrl+Alt+Cmd+u).

```
- (void)invokeTest
{
  for (int i = 0; i < 10000; i++) {
    NSLog(@"***Invocation #%d***", i);
    [super invokeTest];
    [NSRunLoop.mainRunLoop runMode:NSDefaultRunLoopMode beforeDate:NSDate.distantPast];
  }
}
```

**Note:** I also had to change the `dispatch_once_t`s away from `static` so that they would fire again on next iteration.

I then turned on verbose logging `ASEnableVerboseLogging`.

Then I set a breakpoint where `OCMReportFailure` is called. This is important, because the "Test failure breakpoint" Xcode provides only handles the failure macro, not explicitly invoking `recordFailureAtLine:`. We could modify/fork OCMock to use the `_XCTFailureHandler` macro to cover this base in the future.

Then I walked away and got a coffee while the 10000 iterations ran.

When I came back, I was at the failure breakpoint (the extra layout pass wasn't running). So I dumped the log for that iteration, plus the previous iteration. I use Sublime text to strip out all timestamps and pointer values from the log, then I ran `diff` on the two log chunks. The diff showed that in the test fail case, the node still had a pending layout that it wanted to use in `__layout` and in the common case, the node had null pending layout.

I added some more log statements to cover when _pendingDisplayNodeLayout changed and ran the process again. I grabbed the good & bad log chunks, stripped them down, ran the diff and got this diff:

screen shot 2017-07-07 at 7 14 42 pm

**Conclusion:** If CA runs a layout pass on the main thread before the transition can get a hold of nodeC's lock on the background thread, then no extra layout happens because the pending layout that gets established can be used in the layout pass for the new fixture (C's layout between the two target layouts is the same, D's is not and that's why its behavior is more predictable). If instead CA runs its layout pass after the bg layout gets set as pending, nodeC will discard it as part of the layout pass and have to regenerate it. Note that it's really nodeA's lock that matters here, since nodeA's lock is held throughout the calculation process, and nodeA needs layout so it's a race for the two threads to lock node A.

Not sure yet what I'll do with this knowledge, but there it is.

File: Tests/Common/debugbreak.h (hunk 5 above)
Comment (nguyenhuy) [inline Tests/Common/debugbreak.h]: That's sad :(

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by nguyenhuy** [comment]
@Adlai-Holler Please fix the file conflicts and merge away.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/54a87cdd1387e4d745c95de4cbe8cb5c/raw/fa816df493932aca997c2e32c30dd623df6828d1/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/953abfbf5370919bb422522acd0cfa6b/raw/bceb0801054199189c087bd9e3500789323c0566/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/d671480e65bd677124d841e086a6851a/raw/ace485b080f1cd61f8dca5e7b4cdbc3128ed1f68/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/fd0e9f0b767fab0183b2e81cd8d401d5/raw/42a0103c7d1055a30333d5ac736810e9097e490a/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/de6fb4c4251a0a01ad0d5f533aa501db/raw/ad342d6879f0820764eb1ccfe42545dc9cf2551e/log.txt)

File: Source/Layout/ASLayout.mm [hunk 7]
```diff
@@ -373,22 +374,6 @@ - (NSString *)_recursiveDescriptionForLayout:(ASLayout *)layout level:(NSUIntege
...
-  if (ASLayout *layout = ASDynamicCast(object, ASLayout)) {
-    return CGSizeEqualToSize(self.size, layout.size)
-    && ((ASPointIsNull(self.position) && ASPointIsNull(layout.position))
-      || CGPointEqualToPoint(self.position, layout.position))
-    && self.layoutElement == layout.layoutElement
-    && ASObjectIsEqual(self.sublayouts, layout.sublayouts);
-  }
-  return NO;
-}
-
```
Comment (Adlai-Holler) [inline Source/Layout/ASLayout.mm]: Oh wow. Very very scary that this never warned of duplicate method implementation.

**Comment by nguyenhuy** [comment]
FYI, These new tests pass locally for me if I apply this diff on top of #695.

And yes, we should make sure the log has enough info.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/012cb7959926d9f798696e897d3d1e6b/raw/e52ddee3b651cac7a1ba02035951e3d49d433c49/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Pull request merged**